### DISCUSSION TITLE
Common structs

### DISCUSSION THREAD

**Discussion by ambitslix** [protobuf-c/protobuf-c#666] [Q&A]
What is the best way to solve this common problem:

We have runtime structs in our program with dynamically allocated memory:

``` c
struct runtime{
 int32_t unit;
 char msg[64];
 char *dynamic1, *dynamic 2;
}
```

But now we want PB to serialize the runtime struct but we only need unit and msg for packing and transmission, we don't want to pack dynamicx pointers.

We could create separate struct with PB .proto :

``` c
struct packruntime{
 int32_t unit;
 char msg[64];
}
```

Then do something like this in C code:

``` c
struct runtime{
 struct packruntime pr;
 char *dynamic1, *dynamic 2;
}
```

Or we have to copy part of our runtime struct each time into a PB generated packed struct.

Any suggestions?

**Comment by lipnitsk** [comment]
If you weren't using strings, I would suggest to define the entire struct in .proto and define elements you don't want to pack as `optional`. That way, you could use the has_ fields to control what needs to get packed. Unfortunately, char arrays key off of the pointer being NULL on whether or not they will be packed. So in your specific case I would probably use your second suggestion, to avoid struct duplication.
